dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1104 @@
+    auto max   = stencil.max();
+
+    if( _specs[index].extent() == 0 )
+      ++_num_regions;
+
+    if( max > _specs[index].extent() )
+      _specs[index] = HaloRegionSpecT( index, max );
+  }
+
+  bool nextRegion( const StencilT& stencil, StencilT& stencil_combination ) {
+    for( auto d = 0; d < NumDimensions; ++d ) {
+      if( stencil[d] == 0 )
+        continue;
+      stencil_combination[d] = ( stencil_combination[d] == 0 ) ? stencil[d] : 0;
+      if( stencil_combination[d] == 0 ) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+ private:
+  SpecsT      _specs{};
+  std::size_t _num_regions{ 0 };
+};  // HaloSpec
+
+template <typename ElementT, typename PatternT,
+          typename PointerT   = GlobPtr<ElementT, PatternT>,
+          typename ReferenceT = GlobRef<ElementT>>
+class RegionIter
+: public std::iterator<std::random_access_iterator_tag, ElementT,
+                       typename PatternT::index_type, PointerT, ReferenceT> {
+ private:
+  using SelfT = RegionIter<ElementT, PatternT, PointerT, ReferenceT>;
+  using GlobMemT =
+    GlobStaticMem<ElementT, dash::allocator::SymmetricAllocator<ElementT>>;
+  using ViewSpecT = typename PatternT::viewspec_type;
+
+  static const dim_t NumDimensions = PatternT::ndim();
+
+ public:
+  using index_type = typename PatternT::index_type;
+  using size_type  = typename PatternT::size_type;
+
+  using reference       = ReferenceT;
+  using const_reference = const reference;
+  using pointer         = PointerT;
+  using const_pointer   = const pointer;
+
+ public:
+  /**
+   * Constructor, creates a block boundary iterator on multiple boundary
+   * regions.
+   */
+  RegionIter( GlobMemT& globmem, const PatternT& pattern,
+              const ViewSpecT& _region_view, index_type pos, size_type size )
+  : _globmem( globmem ), _pattern( pattern ), _region_view( _region_view ),
+    _idx( pos ), _max_idx( size - 1 ), _myid( pattern.team().myid() ),
+    _lbegin( globmem.lbegin() ) {}
+
+  /**
+   * Copy constructor.
+   */
+  RegionIter( const SelfT& other ) = default;
+
+  /**
+   * Move constructor
+   */
+  RegionIter( SelfT&& other ) = default;
+
+  /**
+   * Assignment operator.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  SelfT& operator=( const SelfT& other ) = default;
+
+  /**
+   * Move assignment operator
+   */
+  SelfT& operator=( SelfT&& other ) = default;
+
+  /**
+   * The number of dimensions of the iterator's underlying pattern.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  static constexpr dim_t ndim() { return NumDimensions; }
+
+  /**
+   * Dereference operator.
+   *
+   * \return  A global reference to the element at the iterator's position.
+   */
+  reference operator*() const { return operator[]( _idx ); }
+
+  /**
+   * Subscript operator, returns global reference to element at given
+   * global index.
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  reference operator[]( index_type idx ) const {
```

Comment:
I'm a bit puzzled by the semantics of `operator[]` on an iterator: My understanding is that the iterator points to an element in a container. This implementation ignores the current position completely. If it's not really needed, I'd suggest to remove it.

Reply:
We are using the subscript operator in other iterators the same way. The standard library uses this operator for the random access iterator different -> http://en.cppreference.com/w/cpp/concept/RandomAccessIterator

Reply:
`*(i + n)` seems to be reasonable if you think in pointer arithmetics. If we use the subscript operator in other places the same way those places need fixing as well. That's all but intuitive!